lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Shared.hs

```diff
@@ -0,0 +1,311 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+-- |
+-- Copyright: © 2018-2021 IOHK
+-- License: Apache-2.0
+--
+-- An implementation of shared script state using
+-- scheme specified in CIP-1854 Multi-signature Wallets.
+
+module Cardano.Wallet.Primitive.AddressDiscovery.Shared
+    (
+    -- ** State
+      SharedState (..)
+    , unsafePendingSharedState
+    , newSharedState
+    , addCosignerAccXPub
+    , purposeCIP1854
+    , isShared
+    ) where
+
+import Prelude
+
+import Cardano.Address.Script
+    ( Cosigner (..)
+    , KeyHash (..)
+    , Script (..)
+    , ScriptTemplate (..)
+    , ValidationLevel (..)
+    , foldScript
+    , validateScriptTemplate
+    )
+import Cardano.Crypto.Wallet
+    ( XPub )
+import Cardano.Wallet.Primitive.AddressDerivation
+    ( Depth (..)
+    , DerivationPrefix (..)
+    , DerivationType (..)
+    , Index (..)
+    , MkKeyFingerprint (..)
+    , NetworkDiscriminant (..)
+    , Role (..)
+    , SoftDerivation
+    , WalletKey (..)
+    )
+import Cardano.Wallet.Primitive.AddressDiscovery
+    ( coinTypeAda )
+import Cardano.Wallet.Primitive.AddressDiscovery.Script
+    ( keyHashFromAccXPubIx )
+import Cardano.Wallet.Primitive.AddressDiscovery.Sequential
+    ( AddressPool
+    , AddressPoolGap
+    , ParentContext (..)
+    , context
+    , lookupAddress
+    , mkAddressPool
+    )
+import Cardano.Wallet.Primitive.Types.Address
+    ( Address (..), AddressState (..) )
+import Control.DeepSeq
+    ( NFData )
+import Data.Coerce
+    ( coerce )
+import Data.Either
+    ( isRight )
+import Data.Proxy
+    ( Proxy (..) )
+import GHC.Generics
+    ( Generic )
+import Type.Reflection
+    ( Typeable )
+
+import qualified Data.Map.Strict as Map
+
+{-------------------------------------------------------------------------------
+                                Shared State
+-------------------------------------------------------------------------------}
+
+-- | Shared wallet is a new kind of wallet owned by one or more co-signers.
+-- | In this type of the wallet, addresses are defined by two monetary scripts
+-- | (one for the ownership of assets and the other one for the ownership of stake).
+-- | The two scripts can be the same, different, also the script for the ownership of stake
+-- | can be absent.
+--
+-- | The shared wallet can be in two states: pending and ready. In a pending state the wallet
```

Comment:
"pending or ready" I imagine, unless we have Schrödinger states :p

Reply:
corrected in https://github.com/input-output-hk/cardano-wallet/pull/2508/commits/c0c35b7f84b3e2863bb52d2fc70e1cfd7eb785d2